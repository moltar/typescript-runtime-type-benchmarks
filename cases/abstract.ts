import clone from 'clone';
import { Data } from '../data';

export abstract class Case implements Case {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protected readonly data: any;

  /**
   * Human readable case name.
   */
  abstract get name(): string;

  constructor(data: Data) {
    this.data = Object.freeze(clone(data));
  }

  /**
   * Validation implementation method.
   *
   * Method returns a `Data` object or throws when invalid data is provided.
   */
  abstract validate(): PromiseLike<Data> | Data;
}
