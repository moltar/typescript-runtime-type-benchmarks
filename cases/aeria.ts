import type { ObjectProperty } from '@aeriajs/types'
import { silentValidator } from '@aeriajs/validation'
import { createCase } from '../benchmarks';

const schema = <const>{
  type: 'object',
  properties: {
    number: {
      type: 'number',
    },
    negNumber: {
      type: 'number',
    },
    maxNumber: {
      type: 'number',
    },
    string: {
      type: 'string',
    },
    longString: {
      type: 'string',
    },
    boolean: {
      type: 'boolean',
    },
    deeplyNested: {
      type: 'object',
      properties: {
        foo: {
          type: 'string',
        },
        num: {
          type: 'number',
        },
        bool: {
          type: 'boolean',
        },
      },
      required: ['foo', 'num', 'bool'],
    },
  },
  required: [
    'number',
    'negNumber',
    'maxNumber',
    'string',
    'longString',
    'boolean',
    'deeplyNested',
  ],
} satisfies ObjectProperty

createCase('aeria', 'parseSafe', () => {
  const [_, validate] = silentValidator(schema, {
    extraneous: true,
    filterOutExtraneous: true,
    coerce: true
  })

  return (data: any) => {
    return validate(data)
  }
})

createCase('aeria', 'parseStrict', () => {
  const [_, validate] = silentValidator(schema, {
    coerce: true
  })

  return (data: any) => {
    return validate(data)
  }
})

createCase('aeria', 'assertLoose', () => {
  const [_, validate] = silentValidator(schema, {
    extraneous: true
  })

  return (data: any) => {
    return validate(data)
  }
})

createCase('aeria', 'assertStrict', () => {
  const [_, validate] = silentValidator(schema)

  return (data: any) => {
    return validate(data)
  }
})
