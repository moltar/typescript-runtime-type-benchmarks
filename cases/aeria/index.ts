import { validate } from '@aeriajs/validation';
import { createCase } from '../../benchmarks';

const schema = {
  type: 'object',
  properties: {
    number: {
      type: 'number',
    },
    negNumber: {
      type: 'number',
    },
    maxNumber: {
      type: 'number',
    },
    string: {
      type: 'string',
    },
    longString: {
      type: 'string',
    },
    boolean: {
      type: 'boolean',
    },
    deeplyNested: {
      type: 'object',
      properties: {
        foo: {
          type: 'string',
        },
        num: {
          type: 'number',
        },
        bool: {
          type: 'boolean',
        },
      },
      required: ['foo', 'num', 'bool'],
    },
  },
  required: [
    'number',
    'negNumber',
    'maxNumber',
    'string',
    'longString',
    'boolean',
    'deeplyNested',
  ],
} as const;

createCase('aeria', 'parseSafe', () => {
  return (data: unknown) => {
    return validate(data, schema, {
      throwOnError: true,
      tolerateExtraneous: true,
    }).result;
  };
});

createCase('aeria', 'parseStrict', () => {
  return (data: unknown) => {
    return validate(data, schema, {
      throwOnError: true,
    }).result;
  };
});

createCase('aeria', 'assertLoose', () => {
  return (data: unknown) => {
    validate(data, schema, {
      throwOnError: true,
      tolerateExtraneous: true,
    });

    return true;
  };
});

createCase('aeria', 'assertStrict', () => {
  return (data: unknown) => {
    validate(data, schema, {
      throwOnError: true,
    });

    return true;
  };
});
