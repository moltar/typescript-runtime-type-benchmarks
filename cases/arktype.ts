import { ArkErrors, type } from 'arktype';
import { createCase } from '../benchmarks';

// As of version 2.1.9, the test fails because of a mutation attempt
// createCase('arktype', 'parseSafe', () => {
//   const t = type({
//     '+': 'delete',
//     number: 'number',
//     negNumber: 'number',
//     maxNumber: 'number',
//     string: 'string',
//     longString: 'string',
//     boolean: 'boolean',
//     deeplyNested: {
//       '+': 'delete',
//       foo: 'string',
//       num: 'number',
//       bool: 'boolean',
//     },
//   });
//   return data => {
//     const r = t(data);
//     if (r instanceof ArkErrors) throw r;
//     return r;
//   };
// });

createCase('arktype', 'parseStrict', () => {
  const t = type({
    '+': 'reject',
    number: 'number',
    negNumber: 'number',
    maxNumber: 'number',
    string: 'string',
    longString: 'string',
    boolean: 'boolean',
    deeplyNested: {
      '+': 'reject',
      foo: 'string',
      num: 'number',
      bool: 'boolean',
    },
  });
  return data => {
    const r = t(data);
    if (r instanceof ArkErrors) throw r;
    return r;
  };
});

createCase('arktype', 'assertLoose', () => {
  const t = type({
    number: 'number',
    negNumber: 'number',
    maxNumber: 'number',
    string: 'string',
    longString: 'string',
    boolean: 'boolean',
    deeplyNested: {
      foo: 'string',
      num: 'number',
      bool: 'boolean',
    },
  });

  return data => {
    if (t.allows(data)) return true;
    throw new Error('Invalid');
  };
});

createCase('arktype', 'assertStrict', () => {
  const t = type({
    '+': 'reject',
    number: 'number',
    negNumber: 'number',
    maxNumber: 'number',
    string: 'string',
    longString: 'string',
    boolean: 'boolean',
    deeplyNested: {
      '+': 'reject',
      foo: 'string',
      num: 'number',
      bool: 'boolean',
    },
  });

  return data => {
    if (t.allows(data)) return true;
    throw new Error('Invalid');
  };
});
