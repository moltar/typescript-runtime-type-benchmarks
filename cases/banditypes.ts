import { boolean, number, object, string } from 'banditypes';
import { addCase } from '../benchmarks';

const dataTypeSafe = object({
  number: number(),
  negNumber: number(),
  maxNumber: number(),
  string: string(),
  longString: string(),
  boolean: boolean(),
  deeplyNested: object({
    foo: string(),
    num: number(),
    bool: boolean(),
  }),
});

addCase('banditypes', 'parseSafe', data => {
  return dataTypeSafe(data);
});

addCase('banditypes', 'assertLoose', data => {
  dataTypeSafe(data);

  return true;
});
