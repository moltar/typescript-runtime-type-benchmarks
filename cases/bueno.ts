import {
  boolean,
  check,
  enUS,
  number,
  object,
  objectExact,
  objectInexact,
  result,
  string,
} from 'bueno';
import { type UnknownData, addCase } from '../benchmarks/index.ts';

const dataType = object({
  number: number,
  negNumber: number,
  maxNumber: number,
  string: string,
  longString: string,
  boolean: boolean,
  deeplyNested: object({
    foo: string,
    num: number,
    bool: boolean,
  }),
});

const dataTypeStrict = objectExact({
  number: number,
  negNumber: number,
  maxNumber: number,
  string: string,
  longString: string,
  boolean: boolean,
  deeplyNested: objectExact({
    foo: string,
    num: number,
    bool: boolean,
  }),
});

const dataTypeLoose = objectInexact({
  number: number,
  negNumber: number,
  maxNumber: number,
  string: string,
  longString: string,
  boolean: boolean,
  deeplyNested: objectInexact({
    foo: string,
    num: number,
    bool: boolean,
  }),
});

addCase('bueno', 'parseSafe', (data: UnknownData) => {
  const err = check(data, dataType, enUS);

  if (err) {
    throw new Error(err);
  }

  return result(data, dataType);
});

addCase('bueno', 'parseStrict', (data: UnknownData) => {
  const err = check(data, dataTypeStrict, enUS);

  if (err) {
    throw new Error(err);
  }

  return result(data, dataTypeStrict);
});

addCase('bueno', 'assertLoose', (data: UnknownData) => {
  const err = check(data, dataTypeLoose, enUS);

  if (err) {
    throw new Error(err);
  }

  return true;
});

addCase('bueno', 'assertStrict', (data: UnknownData) => {
  const err = check(data, dataTypeStrict, enUS);

  if (err) {
    throw new Error(err);
  }

  return true;
});
