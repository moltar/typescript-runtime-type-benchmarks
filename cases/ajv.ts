import Ajv from 'ajv';
import { Case } from './abstract';
import { Data } from '../data';

const ajv = new Ajv();
const schema = {
  $id: 'AjvTest',
  $schema: 'http://json-schema.org/draft-07/schema#',
  type: 'object',
  properties: {
    number: {
      type: 'number',
    },
    negNumber: {
      type: 'number',
    },
    maxNumber: {
      type: 'number',
    },
    string: {
      type: 'string',
    },
    longString: {
      type: 'string',
    },
    boolean: {
      type: 'boolean',
    },
    deeplyNested: {
      type: 'object',
      properties: {
        foo: {
          type: 'string',
        },
        num: {
          type: 'number',
        },
        bool: {
          type: 'boolean',
        },
      },
      required: ['foo', 'num', 'bool'],
    },
  },
  required: [
    'number',
    'negNumber',
    'maxNumber',
    'string',
    'longString',
    'boolean',
    'deeplyNested',
  ],
};
const validate = ajv.compile(schema);

export class AjvCase extends Case implements Case {
  name = 'ajv';

  validate() {
    if (!validate(this.data)) {
      throw new Error(JSON.stringify(ajv.errors, null, 4));
    }
    return this.data as Data;
  }
}
