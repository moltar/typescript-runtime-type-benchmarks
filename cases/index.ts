import { BuenoCase } from './bueno';
import {
  ClassValidatorAsyncCase,
  ClassValidatorSyncCase,
} from './class-validator';
import { ComputedTypesCase } from './computed-types';
import { DecodersCase } from './decoders';
import { IoTsCase } from './io-ts';
import { JointzCase } from './jointz';
import { JsonDecoderCase } from './json-decoder';
import { MarshalCase } from './marshal';
import { MyzodCase } from './myzod';
import { PurifyCase } from './purify-ts';
import { RulrCase } from './rulr';
import { RuntypesCase } from './runtypes';
import { SimpleRuntypesCase } from './simple-runtypes';
import { SuperstructCase } from './superstruct';
import { SuretypeCase } from './suretype';
import { ToiCase } from './toi';
import { TsJsonValidatorCase } from './ts-json-validator';
import { TsUtilsCase } from './ts-utils';
import { TypeOfWebSchemaCase } from './typeofweb-schema';
import { ValitaCase } from './valita';
import { YupCase } from './yup';
import { ZodCase } from './zod';

export const cases = [
  BuenoCase,
  ClassValidatorAsyncCase,
  ClassValidatorSyncCase,
  ComputedTypesCase,
  DecodersCase,
  IoTsCase,
  JointzCase,
  JsonDecoderCase,
  MarshalCase,
  MyzodCase,
  PurifyCase,
  RulrCase,
  RuntypesCase,
  SimpleRuntypesCase,
  SuperstructCase,
  SuretypeCase,
  ToiCase,
  TsJsonValidatorCase,
  TsUtilsCase,
  TypeOfWebSchemaCase,
  ValitaCase,
  YupCase,
  ZodCase,
];
