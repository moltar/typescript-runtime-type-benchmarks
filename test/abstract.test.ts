// eslint-disable-next-line node/no-unpublished-import
import { expectTypeOf } from 'expect-type';
import clone from 'clone';
import { cases } from '../cases/index';
import { DATA } from '../data';

const caseClasses = Object.values(cases);

type CaseTuple = [string, typeof caseClasses[number]];

const caseTuples = caseClasses.map<CaseTuple>(caseClass => {
  return [new caseClass(DATA).name, caseClass];
});

const clonedData = () => clone({ ...DATA });

describe.each(caseTuples)('Case Class: %s', (_caseName, caseClass) => {
  let data = clonedData();

  beforeEach(() => {
    data = clonedData();
  });

  it('should validate type', async () => {
    expect.assertions(1);

    const c = new caseClass(data);

    const res = await c.validate();

    expectTypeOf(res).toMatchTypeOf(DATA);

    expect(c).toBeTruthy();
  });

  it('should return a validated data object', async () => {
    const c = new caseClass(data);
    expect(await c.validate()).toEqual(DATA);
  });

  it('should fail validation when type is wrong', async () => {
    expect.assertions(2);

    Object.assign(data, { boolean: 'foo' });

    const c = new caseClass(data);

    let err;

    try {
      await c.validate();
    } catch (e) {
      err = e;
    }

    expect(err).toBeTruthy();
    expect(err).toMatchSnapshot();
  });

  it('should fail validation when deeply nested type is wrong', async () => {
    expect.assertions(2);

    Object.assign(data.deeplyNested, { bool: 'foo' });

    const c = new caseClass(data);

    let err;

    try {
      await c.validate();
    } catch (e) {
      err = e;
    }

    expect(err).toBeTruthy();
    expect(err).toMatchSnapshot();
  });

  it('should fail validation when prop is missing', async () => {
    expect.assertions(2);

    // @ts-expect-error fixes "the operand of a 'delete' operator must be optional" error
    delete data.boolean;

    const c = new caseClass(data);

    let err;

    try {
      await c.validate();
    } catch (e) {
      err = e;
    }

    expect(err).toBeTruthy();
    expect(err).toMatchSnapshot();
  });
});
