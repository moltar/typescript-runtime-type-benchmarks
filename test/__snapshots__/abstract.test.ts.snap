// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Case Class: class-transformer-validator-async should fail validation when deeply nested type is wrong 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [],
        "constraints": Object {
          "isBoolean": "bool must be a boolean value",
        },
        "property": "bool",
        "target": DeeplyNestedType {
          "bool": "foo",
          "foo": "bar",
          "num": 1,
        },
        "value": "foo",
      },
    ],
    "property": "deeplyNested",
    "target": DataType {
      "boolean": true,
      "deeplyNested": DeeplyNestedType {
        "bool": "foo",
        "foo": "bar",
        "num": 1,
      },
      "longString": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Vivendum intellegat et qui, ei denique consequuntur vix. Semper aeterno percipit ut his, sea ex utinam referrentur repudiandae. No epicuri hendrerit consetetur sit, sit dicta adipiscing ex, in facete detracto deterruisset duo. Quot populo ad qui. Sit fugit nostrum et. Ad per diam dicant interesset, lorem iusto sensibus ut sed. No dicam aperiam vis. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Ex nam agam veri, dicunt efficiantur ad qui, ad legere adversarium sit. Commune platonem mel id, brute adipiscing duo an. Vivendum intellegat et qui, ei denique consequuntur vix. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      "maxNumber": 1.7976931348623157e+308,
      "negNumber": -1,
      "number": 1,
      "string": "string",
    },
    "value": DeeplyNestedType {
      "bool": "foo",
      "foo": "bar",
      "num": 1,
    },
  },
]
`;

exports[`Case Class: class-transformer-validator-async should fail validation when prop is missing 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "boolean must be a boolean value",
    },
    "property": "boolean",
    "target": DataType {
      "deeplyNested": DeeplyNestedType {
        "bool": false,
        "foo": "bar",
        "num": 1,
      },
      "longString": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Vivendum intellegat et qui, ei denique consequuntur vix. Semper aeterno percipit ut his, sea ex utinam referrentur repudiandae. No epicuri hendrerit consetetur sit, sit dicta adipiscing ex, in facete detracto deterruisset duo. Quot populo ad qui. Sit fugit nostrum et. Ad per diam dicant interesset, lorem iusto sensibus ut sed. No dicam aperiam vis. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Ex nam agam veri, dicunt efficiantur ad qui, ad legere adversarium sit. Commune platonem mel id, brute adipiscing duo an. Vivendum intellegat et qui, ei denique consequuntur vix. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      "maxNumber": 1.7976931348623157e+308,
      "negNumber": -1,
      "number": 1,
      "string": "string",
    },
    "value": undefined,
  },
]
`;

exports[`Case Class: class-transformer-validator-async should fail validation when type is wrong 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "boolean must be a boolean value",
    },
    "property": "boolean",
    "target": DataType {
      "boolean": "foo",
      "deeplyNested": DeeplyNestedType {
        "bool": false,
        "foo": "bar",
        "num": 1,
      },
      "longString": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Vivendum intellegat et qui, ei denique consequuntur vix. Semper aeterno percipit ut his, sea ex utinam referrentur repudiandae. No epicuri hendrerit consetetur sit, sit dicta adipiscing ex, in facete detracto deterruisset duo. Quot populo ad qui. Sit fugit nostrum et. Ad per diam dicant interesset, lorem iusto sensibus ut sed. No dicam aperiam vis. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Ex nam agam veri, dicunt efficiantur ad qui, ad legere adversarium sit. Commune platonem mel id, brute adipiscing duo an. Vivendum intellegat et qui, ei denique consequuntur vix. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      "maxNumber": 1.7976931348623157e+308,
      "negNumber": -1,
      "number": 1,
      "string": "string",
    },
    "value": "foo",
  },
]
`;

exports[`Case Class: class-transformer-validator-sync should fail validation when deeply nested type is wrong 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [],
        "constraints": Object {
          "isBoolean": "bool must be a boolean value",
        },
        "property": "bool",
        "target": DeeplyNestedType {
          "bool": "foo",
          "foo": "bar",
          "num": 1,
        },
        "value": "foo",
      },
    ],
    "property": "deeplyNested",
    "target": DataType {
      "boolean": true,
      "deeplyNested": DeeplyNestedType {
        "bool": "foo",
        "foo": "bar",
        "num": 1,
      },
      "longString": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Vivendum intellegat et qui, ei denique consequuntur vix. Semper aeterno percipit ut his, sea ex utinam referrentur repudiandae. No epicuri hendrerit consetetur sit, sit dicta adipiscing ex, in facete detracto deterruisset duo. Quot populo ad qui. Sit fugit nostrum et. Ad per diam dicant interesset, lorem iusto sensibus ut sed. No dicam aperiam vis. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Ex nam agam veri, dicunt efficiantur ad qui, ad legere adversarium sit. Commune platonem mel id, brute adipiscing duo an. Vivendum intellegat et qui, ei denique consequuntur vix. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      "maxNumber": 1.7976931348623157e+308,
      "negNumber": -1,
      "number": 1,
      "string": "string",
    },
    "value": DeeplyNestedType {
      "bool": "foo",
      "foo": "bar",
      "num": 1,
    },
  },
]
`;

exports[`Case Class: class-transformer-validator-sync should fail validation when prop is missing 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "boolean must be a boolean value",
    },
    "property": "boolean",
    "target": DataType {
      "deeplyNested": DeeplyNestedType {
        "bool": false,
        "foo": "bar",
        "num": 1,
      },
      "longString": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Vivendum intellegat et qui, ei denique consequuntur vix. Semper aeterno percipit ut his, sea ex utinam referrentur repudiandae. No epicuri hendrerit consetetur sit, sit dicta adipiscing ex, in facete detracto deterruisset duo. Quot populo ad qui. Sit fugit nostrum et. Ad per diam dicant interesset, lorem iusto sensibus ut sed. No dicam aperiam vis. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Ex nam agam veri, dicunt efficiantur ad qui, ad legere adversarium sit. Commune platonem mel id, brute adipiscing duo an. Vivendum intellegat et qui, ei denique consequuntur vix. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      "maxNumber": 1.7976931348623157e+308,
      "negNumber": -1,
      "number": 1,
      "string": "string",
    },
    "value": undefined,
  },
]
`;

exports[`Case Class: class-transformer-validator-sync should fail validation when type is wrong 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "boolean must be a boolean value",
    },
    "property": "boolean",
    "target": DataType {
      "boolean": "foo",
      "deeplyNested": DeeplyNestedType {
        "bool": false,
        "foo": "bar",
        "num": 1,
      },
      "longString": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Vivendum intellegat et qui, ei denique consequuntur vix. Semper aeterno percipit ut his, sea ex utinam referrentur repudiandae. No epicuri hendrerit consetetur sit, sit dicta adipiscing ex, in facete detracto deterruisset duo. Quot populo ad qui. Sit fugit nostrum et. Ad per diam dicant interesset, lorem iusto sensibus ut sed. No dicam aperiam vis. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Ex nam agam veri, dicunt efficiantur ad qui, ad legere adversarium sit. Commune platonem mel id, brute adipiscing duo an. Vivendum intellegat et qui, ei denique consequuntur vix. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      "maxNumber": 1.7976931348623157e+308,
      "negNumber": -1,
      "number": 1,
      "string": "string",
    },
    "value": "foo",
  },
]
`;

exports[`Case Class: computed-types should fail validation when deeply nested type is wrong 1`] = `[TypeError: deeplyNested: deeplyNested.bool: Expect value to be "boolean"]`;

exports[`Case Class: computed-types should fail validation when prop is missing 1`] = `[TypeError: boolean: Expect value to be "boolean"]`;

exports[`Case Class: computed-types should fail validation when type is wrong 1`] = `[TypeError: boolean: Expect value to be "boolean"]`;

exports[`Case Class: io-ts should fail validation when deeply nested type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: io-ts should fail validation when prop is missing 1`] = `[Error: Invalid]`;

exports[`Case Class: io-ts should fail validation when type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: jointz should fail validation when deeply nested type is wrong 1`] = `
Array [
  Object {
    "message": "must be one of true, false",
    "path": Array [
      "deeplyNested",
      "bool",
    ],
    "value": "foo",
  },
]
`;

exports[`Case Class: jointz should fail validation when prop is missing 1`] = `
Array [
  Object {
    "message": "required key \\"boolean\\" was not defined",
    "path": Array [],
    "value": Object {
      "deeplyNested": Object {
        "bool": false,
        "foo": "bar",
        "num": 1,
      },
      "longString": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Vivendum intellegat et qui, ei denique consequuntur vix. Semper aeterno percipit ut his, sea ex utinam referrentur repudiandae. No epicuri hendrerit consetetur sit, sit dicta adipiscing ex, in facete detracto deterruisset duo. Quot populo ad qui. Sit fugit nostrum et. Ad per diam dicant interesset, lorem iusto sensibus ut sed. No dicam aperiam vis. Pri posse graeco definitiones cu, id eam populo quaestio adipiscing, usu quod malorum te. Ex nam agam veri, dicunt efficiantur ad qui, ad legere adversarium sit. Commune platonem mel id, brute adipiscing duo an. Vivendum intellegat et qui, ei denique consequuntur vix. Offendit eleifend moderatius ex vix, quem odio mazim et qui, purto expetendis cotidieque quo cu, veri persius vituperata ei nec. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      "maxNumber": 1.7976931348623157e+308,
      "negNumber": -1,
      "number": 1,
      "string": "string",
    },
  },
]
`;

exports[`Case Class: jointz should fail validation when type is wrong 1`] = `
Array [
  Object {
    "message": "must be one of true, false",
    "path": Array [
      "boolean",
    ],
    "value": "foo",
  },
]
`;

exports[`Case Class: json-decoder should fail validation when deeply nested type is wrong 1`] = `[Error: deeplyNested: bool: expected boolean, got string]`;

exports[`Case Class: json-decoder should fail validation when prop is missing 1`] = `[Error: boolean: expected boolean, got undefined]`;

exports[`Case Class: json-decoder should fail validation when type is wrong 1`] = `[Error: boolean: expected boolean, got string]`;

exports[`Case Class: marshal should fail validation when deeply nested type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: marshal should fail validation when prop is missing 1`] = `[Error: Invalid]`;

exports[`Case Class: marshal should fail validation when type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: myzod should fail validation when deeply nested type is wrong 1`] = `[MyZodError: error parsing object at path: "deeplyNested.bool" - expected type to be boolean but got string]`;

exports[`Case Class: myzod should fail validation when prop is missing 1`] = `[MyZodError: error parsing object at path: "boolean" - expected type to be boolean but got undefined]`;

exports[`Case Class: myzod should fail validation when type is wrong 1`] = `[MyZodError: error parsing object at path: "boolean" - expected type to be boolean but got string]`;

exports[`Case Class: purify-ts should fail validation when deeply nested type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: purify-ts should fail validation when prop is missing 1`] = `[Error: Invalid]`;

exports[`Case Class: purify-ts should fail validation when type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: quartet should fail validation when deeply nested type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: quartet should fail validation when prop is missing 1`] = `[Error: Invalid]`;

exports[`Case Class: quartet should fail validation when type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: rulr should fail validation when deeply nested type is wrong 1`] = `[ValidationErrors: deeplyNested.bool: expected boolean]`;

exports[`Case Class: rulr should fail validation when prop is missing 1`] = `[ValidationErrors: boolean: expected boolean]`;

exports[`Case Class: rulr should fail validation when type is wrong 1`] = `[ValidationErrors: boolean: expected boolean]`;

exports[`Case Class: runtypes should fail validation when deeply nested type is wrong 1`] = `[ValidationError: Expected boolean, but was string]`;

exports[`Case Class: runtypes should fail validation when prop is missing 1`] = `[ValidationError: Expected boolean, but was undefined]`;

exports[`Case Class: runtypes should fail validation when type is wrong 1`] = `[ValidationError: Expected boolean, but was string]`;

exports[`Case Class: toi should fail validation when deeply nested type is wrong 1`] = `[Error: value does not match structure]`;

exports[`Case Class: toi should fail validation when prop is missing 1`] = `[Error: value does not match structure]`;

exports[`Case Class: toi should fail validation when type is wrong 1`] = `[Error: value does not match structure]`;

exports[`Case Class: ts-json-validator should fail validation when deeply nested type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: ts-json-validator should fail validation when prop is missing 1`] = `[Error: Invalid]`;

exports[`Case Class: ts-json-validator should fail validation when type is wrong 1`] = `[Error: Invalid]`;

exports[`Case Class: ts-utils should fail validation when deeply nested type is wrong 1`] = `
DecodeError {
  "expected": [Function],
  "key": Array [
    TypedObject {
      "name": "Data",
      "type": "TypedObject",
    },
    ObjectKey {
      "name": "deeplyNested",
      "type": "ObjectKey",
    },
    TypedObject {
      "name": "DeeplyNested",
      "type": "TypedObject",
    },
    ObjectKey {
      "name": "bool",
      "type": "ObjectKey",
    },
  ],
  "val": "foo",
}
`;

exports[`Case Class: ts-utils should fail validation when prop is missing 1`] = `
DecodeError {
  "expected": [Function],
  "key": Array [
    TypedObject {
      "name": "Data",
      "type": "TypedObject",
    },
    ObjectKey {
      "name": "boolean",
      "type": "ObjectKey",
    },
  ],
  "val": undefined,
}
`;

exports[`Case Class: ts-utils should fail validation when type is wrong 1`] = `
DecodeError {
  "expected": [Function],
  "key": Array [
    TypedObject {
      "name": "Data",
      "type": "TypedObject",
    },
    ObjectKey {
      "name": "boolean",
      "type": "ObjectKey",
    },
  ],
  "val": "foo",
}
`;

exports[`Case Class: yup should fail validation when deeply nested type is wrong 1`] = `[ValidationError: deeplyNested.bool must be a \`boolean\` type, but the final value was: \`"foo"\`.]`;

exports[`Case Class: yup should fail validation when prop is missing 1`] = `[ValidationError: boolean is a required field]`;

exports[`Case Class: yup should fail validation when type is wrong 1`] = `[ValidationError: boolean must be a \`boolean\` type, but the final value was: \`"foo"\`.]`;

exports[`Case Class: zod should fail validation when deeply nested type is wrong 1`] = `[Error: deeplyNested.bool: Non-boolean type: string]`;

exports[`Case Class: zod should fail validation when prop is missing 1`] = `[Error: boolean: Non-boolean type: undefined]`;

exports[`Case Class: zod should fail validation when type is wrong 1`] = `[Error: boolean: Non-boolean type: string]`;
